runtime/sema/type.go

```diff
@@ -5333,6 +5419,34 @@ func checkSubTypeWithoutEquality(subType Type, superType Type) bool {
 			//
 			// The holder of the reference may not gain more permissions or knowledge.
 
+			return false
+
+		case *InterfaceType:
+			switch typedInnerSubType := typedSubType.Type.(type) {
+			// An unauthorized reference to an interface type `&U`
+			// is a supertype of a reference to a composite type `&T`:
+			// if that composite type's conformance set includes that interface.
+			//
+			// This is equivalent in principle to the check we would perform to check
+			// if `&T <: &{U}`, the singleton restricted set containing only `U`.
+			case *CompositeType:
+				return typedInnerSubType.ExplicitInterfaceConformanceSet().Contains(typedInnerSuperType)
+			// An unauthorized reference to an interface type `&T`
+			// is a supertype of a reference to a restricted type `&{U}`:
+			// if the restriction set contains that explicit interface type.
+			//
+			// Note that this does not check whether the restricted type's restricted type conforms to the interface;
+			// i.e. whether in `&R{X} <: &I`, `R <: I`. This is intentional;
+			// when checking whether an attachment declared for `I` is accessible on a value of type `&R{X}`,
+			// even if `R <: I`, we only want to allow access if `X <: I`
+			//
```

Comment:
Maybe add the great example for the non-reference case you have below here too:


```suggestion
			//
			// This particular case comes up when checking attachment access;
			// resource interface I { /* ... */ }
			// attachment A for I { /* ... */ }
			//
			// let i : &{I} = ... // some operation constructing `i`
			// let a = i[A] // must here check that `i`'s type is a subtype of `A`'s base type, or that &{I} <: &I
```